IotHubConsumer/ReceiverConsumer.cs

```diff
@@ -0,0 +1,87 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Messaging.EventHubs.Consumer;
+using Microsoft.Extensions.Configuration;
+
+namespace IotHubConsumer
+{
+    public class ReceiverConsumer
+    {
+        private readonly IConfiguration configuration;
+        public HashSet<TestElement> MessagesWithRetainSet { get; set; } = new HashSet<TestElement>();
+
+        public ReceiverConsumer(IConfiguration configuration)
+        {
+            this.configuration = configuration;
+        }
+
+        // Asynchronously create a PartitionReceiver for a partition and then start
+        // reading any messages sent from the simulated client.
+        public async Task ReceiveMessagesFromDeviceAsync(CancellationToken cancellationToken)
+        {
+            var connectionString = configuration["EventHubCompatibleEndpoint"];
+            var eventHubName = configuration["EventHubName"];
+            await using var consumer = new EventHubConsumerClient(EventHubConsumerClient.DefaultConsumerGroupName, connectionString, eventHubName);
+
+            Console.WriteLine("Listening for messages on all partitions");
+
+            try
+            {
+                // Begin reading events for all partitions, starting with the first event in each partition and waiting indefinitely for
+                // events to become available.  Reading can be canceled by breaking out of the loop when an event is processed or by
+                // signaling the cancellation token.
+                //
+                // The "ReadEventsAsync" method on the consumer is a good starting point for consuming events for prototypes
+                // and samples.  For real-world production scenarios, it is strongly recommended that you consider using the
+                // "EventProcessorClient" from the "Azure.Messaging.EventHubs.Processor" package.
+                //
+                // More information on the "EventProcessorClient" and its benefits can be found here:
+                //   https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs.Processor/README.md
+
+
+                await foreach (var partitionEvent in consumer.ReadEventsAsync(cancellationToken))
+                {
+                    Console.WriteLine("Message received on partition {0}:", partitionEvent.Partition.PartitionId);
+
+                    var data = Encoding.UTF8.GetString(partitionEvent.Data.Body.ToArray());
+
+                    if (partitionEvent.Data.Properties.ContainsKey("mqtt-retain")) //TODO: can remove this if and add all messages for other unit tests
+                    {
+                        MessagesWithRetainSet.Add(new TestElement { Payload = data, RetainFlag = partitionEvent.Data.Properties["mqtt-retain"].ToString() });
```

Comment:
I'd add the partitionEvent directly into the list, instead of creating a new object for it

Reply:
If we add the eventPartition object to the list, we would have to deserialise its data content when we do the lookup in the tests, which would make it more complex. If we end up using the message list in other tests, we would have the content already stored as a string.